packages/ra-core/src/controller/edit/useEditController.spec.tsx

```diff
@@ -881,4 +881,40 @@ describe('useEditController', () => {
             expect.any(Function)
         );
     });
+
+    it('The save function should return errors from the update call in pessimistic mode', async () => {
+        let post = { id: 12 };
+        const update = jest.fn().mockImplementationOnce(() => {
+            return Promise.reject({ body: { errors: { foo: 'invalid' } } });
+        });
+        const dataProvider = ({
+            getOne: () => Promise.resolve({ data: post }),
+            update,
+        } as unknown) as DataProvider;
+        let saveCallback;
+        render(
+            <CoreAdminContext dataProvider={dataProvider}>
+                <EditController {...defaultProps} mutationMode="pessimistic">
+                    {({ save, record }) => {
+                        saveCallback = save;
+                        return <>{JSON.stringify(record)}</>;
+                    }}
+                </EditController>
+            </CoreAdminContext>
+        );
+        await new Promise(resolve => setTimeout(resolve, 10));
+        screen.getByText('{"id":12}');
```

Comment:
remove the await above, and instead replace this line by

```jsx
await screen.findByText('{"id":12}');
```

That way you'll never wait longer than necessary